### PR TITLE
[SMALLFIX] make masks auto update, and make permission checking respect ACL mask

### PR DESCRIPTION
**Pull request by yuzhu**
Before this change, we were ignoring ACL masks. After this change, we will take ACL masks into consideration when we do permission checking. Also we automatically sets masks when updating other relevant entries.

### CODE REVIEW

**Comment by yuzhu** [comment]
@gpang PTAL.

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

File: core/common/src/main/java/alluxio/security/authorization/AccessControlList.java [hunk 1]
```diff
@@ -237,7 +247,13 @@ public void setEntry(AclEntry entry) {
...
       case NAMED_USER:  // fall through
-      case NAMED_GROUP: // fall through
+      case NAMED_GROUP:
+        if (mExtendedEntries == null) {
+          mExtendedEntries = new ExtendedACLEntries();
+        }
+        mExtendedEntries.setEntry(entry);
+        updateMask();
+        return;
       case MASK:
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authorization/AccessControlList.java]: Does setting the mask directly need to compute anything additional?

File: core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java [hunk 2]
```diff
@@ -206,6 +206,38 @@ public AclActions getMask() {
     return actions;
   }
 
+  /**
+   * Update the mask to be the union of owning group entry, named user entry and named group entry.
+   * @param groupActions the group entry to be integrated into the mask
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java]: why does the entire mask have to be recomputed, when it sounds like we just want to update it with the new group actions?

File: core/common/src/main/java/alluxio/security/authorization/AccessControlList.java (hunk 1 above)
Comment (yuzhu) [reply core/common/src/main/java/alluxio/security/authorization/AccessControlList.java]: nope.

File: core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java (hunk 2 above)
Comment (yuzhu) [reply core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java]: the entire mask needs to be recomputed in the case of a removal of entry. 
I updated the code so that the mask is recomputed once for each command, rather than each addition / removal.

**Comment by yuzhu** [comment]
jenkins, test this please

File: core/common/src/main/java/alluxio/security/authorization/AccessControlList.java [hunk 3]
```diff
@@ -248,6 +258,7 @@ public void setEntry(AclEntry entry) {
         Mode modeOwner = new Mode(mMode);
         modeOwner.setOwnerBits(entry.getActions().toModeBits());
         mMode = modeOwner.toShort();
+
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authorization/AccessControlList.java]: NIT: unnecessary change?

File: core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java (hunk 2 above)
Comment (gpang) [reply core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java]: Can you comment that this must be called when the entries are modified? (And maybe also add comments to setEntries() to call this after changing ACLs)

File: core/server/master/src/main/java/alluxio/master/file/meta/Inode.java [hunk 4]
```diff
@@ -262,6 +263,41 @@ public T replaceAcl(List<AclEntry> entries) {
     return setAcl(entries);
   }
 
+  /**
+   * Update Mask for the Inode.
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/Inode.java]: Comment on when this should be called. Can this just be a part of the update process?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 5]
```diff
@@ -3245,12 +3245,15 @@ private void setAclSingleInode(RpcContext rpcContext, SetAclAction action,
       case REPLACE:
         // fully replace the acl for the path
         targetInode.replaceAcl(entries);
+        targetInode.updateMask(entries);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: why can't updateMask() functionality just be a part of `replaceAcl` or `setAcl`?

File: core/common/src/main/java/alluxio/security/authorization/AccessControlList.java (hunk 3 above)
Comment (yuzhu) [reply core/common/src/main/java/alluxio/security/authorization/AccessControlList.java]: done

File: core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java (hunk 2 above)
Comment (yuzhu) [reply core/common/src/main/java/alluxio/security/authorization/ExtendedACLEntries.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/Inode.java (hunk 4 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/Inode.java]: made it part of the update process.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 5 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: made it part of those functions.

**Comment by yuzhu** [comment]
@gpang addressed comments, PTAL

File: core/common/src/main/java/alluxio/security/authorization/AccessControlList.java [hunk 6]
```diff
@@ -226,15 +226,26 @@ public void setMode(short mode) {
     mMode = mode;
   }
 
+  /**
+   * Update the mask to be the union of owning group entry, named user entry and named group entry.
+   * This method should be called when the aforementioned entries are modified.
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authorization/AccessControlList.java]: This "must" be called, right?

Comment (yuzhu) [reply core/common/src/main/java/alluxio/security/authorization/AccessControlList.java]: done

**Comment by yuzhu** [comment]
@gpang comments addressed. PTAL.
